Lombiq.Hosting.Azure.ApplicationInsights/Extensions/ApplicationInsightsInitializerExtensions.cs

```diff
@@ -14,12 +14,20 @@ namespace Microsoft.Extensions.DependencyInjection;
 public static class ApplicationInsightsInitializerExtensions
 {
     /// <summary>
-    /// Use this extension method to initialize Application Insights. Should be used in the application Program.cs file.
+    /// Initializes Application Insights for Orchard Core. Should be used in the application Program.cs file.
     /// </summary>
     public static IServiceCollection AddOrchardCoreApplicationInsightsTelemetry(
         this IServiceCollection services,
-        ConfigurationManager configurationManager)
+        IConfiguration configurationManager)
     {
+        var connectionString = configurationManager.GetValue<string>("ApplicationInsights:ConnectionString");
```

Comment:
> access the AI configuration object to check if it's configured

Do you mean ApplicationInsightsServiceOptions?

Reply:
No, `ApplicationInsightsServiceOptions`, AI's own config.

Reply:
Didn't I just write that? 🤔

Reply:
Anyway, that won't be configured yet.

And Instrumentation key being set directly is obsolete: https://github.com/microsoft/ApplicationInsights-dotnet/issues/2560

Reply:
Yeah :D.

Yes, and I didn't mean to set it but to get it. Even if it's obsolete, people can configure it and it'll work. So, you need to check both that and the instrumentation key.

You can also bind `IConfiguration` to that instead of directly reading out the config, at least.

Reply:
So instead of this:
![image](https://github.com/Lombiq/Orchard-Azure-Application-Insights/assets/19938236/2b964109-5699-4b21-8658-06dc1e110b91)
We should do this:
![image](https://github.com/Lombiq/Orchard-Azure-Application-Insights/assets/19938236/a1a7ed0e-bbcd-456d-ae45-0d1180f255e4)


Why is the second approach better than the first one?

Reply:
Because the configuration keys and their types are the API of the AI library that you'd otherwise duplicate (and we'd have to maintain if it changes).

BTW for an Azure App Service, the `APPINSIGHTS_INSTRUMENTATIONKEY` and `APPLICATIONINSIGHTS_CONNECTION_STRING` app configurations will be set automatically, these should work too. I don't know where AI reads these, that's why I said the best would be to read the configuration object, i.e. resolve `IConfiguration<ApplicationInsightsServiceOptions>`.

Reply:
AI reads these when we are calling `services.AddApplicationInsightsTelemetry(configurationManager);` a few lines below and to be more specific in the [AddTelemetryConfiguration()](https://github.com/microsoft/ApplicationInsights-dotnet/blob/main/NETCORE/src/Shared/Extensions/ApplicationInsightsExtensions.cs#L234C41-L234C41) internal function.

Reply:
So there's a chicken and egg problem. Check if setting `DeveloperMode` solves our problem. If not, then if we can remove AI after adding it.

Reply:
The current code references the config directly. See what I've written previously. Or we can do the same as when `EnableOfflineOperation`.

Reply:
How is developer mode connected to this? https://learn.microsoft.com/en-us/azure/azure-monitor/app/configuration-with-applicationinsights-config#developer-mode

Reply:
> The current code references the config directly. See what I've written previously. Or we can do the same as when EnableOfflineOperation.

It is a simple binding currently like you wrote "You can also bind IConfiguration to that instead of directly reading out the config, at least.". So we are only depending on "ApplicationInsights" configuration string. I'm not sure what are you referring to now as "what I've written previously", please quote it.

Reply:
So, the goal here is to determine if AI is configured.

- You can do that by checking if any of the connection string or instrumentation key is configured. This might not be the only way though and AI perhaps exposes a way to determine if it's configured.
- The value can come from virtually anywhere; one option is from `IConfiguration`, but neither are single keys the only ones used there (since an Azure App Service, will use the `APPINSIGHTS_INSTRUMENTATIONKEY` and `APPLICATIONINSIGHTS_CONNECTION_STRING` app configurations), nor is a key in `IConfiguration` the only way to set those (`ApplicationInsightsExtensions` has others too, see e.g. [here](https://github.com/microsoft/ApplicationInsights-dotnet/blob/main/NETCORE/src/Shared/Extensions/ApplicationInsightsExtensions.cs#L220)). Because of this, the current code will think there's no configuration when running in an App Service.
- Due to this, the only reliable way is to inject an `IOptions<ApplicationInsightsServiceOptions>` and read out the properties.
- It's up to you how you achieve this, I merely suggested pointers.

Reply:
Pointers?

Reply:
I don't understand the question. Are the goals clear for your here?

Reply:
Yes, but it is not possible to use the `IOptions<ApplicationInsightsServiceOptions>` object here because it is not initiated until the WebApplication is built with `builder.Build();` in _Program.cs_. That's when the `services.Configure(options)` will be run, and thus `ApplicationInsightsServiceOptions` will get its values.

So we either don't care about messages sent to nowhere (that was the original issue here) and I delete these checks.

Or we don't care about the settings here, we add AI, then in the tenant level _Startup.cs_ `ConfigureServices()` we handle the missing connection string and instrumentation key the same as we enable offline operation.

Reply:
You've written that AI configures itself during `AddApplicationInsightsTelemetry()`. So, the configuration should be available after this somewhere, no?

Reply:
For info while building the app we already can get values from the config providers that are already added. I used it in one PR see https://github.com/OrchardCMS/OrchardCore/pull/14550

https://github.com/OrchardCMS/OrchardCore/blob/5a472d483fc360d1e1aa69dc7e098be6a038e62c/src/OrchardCore/OrchardCore.Configuration.KeyVault/Extensions/AzureKeyVaultConfigurationExtension.cs#L27-L33

            builder.ConfigureAppConfiguration((context, builder) =>
            {
                // Here 'builder' is a config manager being a builder and also an 'IConfigurationRoot'
                // if get from the 'context', allowing to get values from the providers already added
                // without having to build a configuration on the fly that would need to be disposed.
                AddOrchardCoreAzureKeyVault(builder, context.Configuration, tokenCredential);
            });

Reply:
> So, the configuration should be available after this somewhere, no?

No, it adds the configuration Action that will be run when services.Configure is called.

Thank you @jtkech I'll check it!

Reply:
@jtkech 
Hmm I think this is the same as we are currently doing. The problem is not that we can't get the values from the already added providers, but to get the values from the instantiated `IOptions<ApplicationInsightsServiceOptions>`, which would only happen on `WebApplicationBuilder.Build();`.